### ISSUE TITLE
Accessing server environment via cli

### ISSUE THREAD

**Issue by maxim-uvarov** [cptpiepmatz/nu-jupyter-kernel#4]
I wonder if it is possible to send commands to the server via a CLI interface and receive responses back via stdout.

I suppose this way of using the server might have several real-life applications. I would like to implement notebook functionality while using the Helix editor to write pure text.

I can write a custom Nushell command that will parse a text document. When this command finds some special tags, it calls the server, waits for a response, and returns the results back.

In Helix, I can just pipe the whole document or a part of it to this custom command, and the result is supposed to augment the document.

**Comment by maxim-uvarov** [comment]
Or in Nushell, I can employ the `watch` command, which can monitor changes in a markdown file containing chunks of Nushell code.

So when I update the document—let's say I add the ` ```nu {update} ` tag to some chunk—the code of this chunk will be sent to the running server, and the response will be used to update the initial markdown, with the removed `{update}` tag in it.

**Comment by cptpiepmatz** [comment]
Check out [jupyter_console](https://github.com/jupyter/jupyter_console). It is a cli implementation of a jupyter client. Maybe that is already what you want. If not, you can check out [jupyter_client](https://github.com/jupyter/jupyter_client) and [nb_client](https://github.com/jupyter/nbclient). Both provide an API to interact with a jupyter kernel. The [docs](https://jupyter-client.readthedocs.io/en/stable/index.html) for jupyter client are also pretty solid.

Since the jupyter API only depents on ZeroMQ communication, you can also implement things in Rust. Some people already started doing things, see [jupyter](https://crates.io/crates/jupyter), [jupyter-client](https://crates.io/crates/jupyter-client) or [jupyter-types](https://crates.io/crates/jupyter-types) on crates.io. I did not use any of them, I just implemented the types accordings to the [messaging docs of the `jupyter_client`](https://jupyter-client.readthedocs.io/en/stable/messaging.html).

**Comment by cptpiepmatz** [comment]
Oh, the jupyter console crashes for me on unknown commands, I'll check what I need to do to fix that:

```
nu ❯ jupyter console --kernel nu
unknown request history_request
could not recv message
C:\Users\derPi\AppData\Roaming\uv\tools\jupyter-core\Lib\site-packages\jupyter_console\ptshell.py:136: UserWarning: No lexer found for language 'nushell'. Treating as plain text.
  warn("No lexer found for language %r. Treating as plain text." % name)
Jupyter console 6.6.3

     __  ,
 .--()°'.' You're running `nu-jupyter-kernel`,
'|, . ,'   based on the `nu` language,
 !_-(_\    where all data is structured!

In [1]: stuff
Error: nu::compile::run_external_not_found

  × External calls are not supported.
   ╭─[cell[1]#1:1:1]
 1 │ stuff
   · ──┬──
   ·   ╰── can't be run in this context
   ╰────
  help: `run-external` was not found in scope

Task exception was never retrieved
future: <Task finished name='Task-6' coro=<ZMQTerminalInteractiveShell.interact() done, defined at C:\Users\derPi\AppData\Roaming\uv\tools\jupyter-core\Lib\site-packages\jupyter_console\ptshell.py:637> exception=KeyError('execution_count')>
Traceback (most recent call last):
  File "C:\Users\derPi\AppData\Roaming\uv\tools\jupyter-core\Lib\site-packages\jupyter_console\ptshell.py", line 650, in interact
    self.run_cell(code, store_history=True)
  File "C:\Users\derPi\AppData\Roaming\uv\tools\jupyter-core\Lib\site-packages\jupyter_console\ptshell.py", line 732, in run_cell
    self.handle_execute_reply(msg_id, timeout=0.05)
  File "C:\Users\derPi\AppData\Roaming\uv\tools\jupyter-core\Lib\site-packages\jupyter_console\ptshell.py", line 771, in handle_execute_reply
    self.execution_count = int(content["execution_count"] + 1)
                               ~~~~~~~^^^^^^^^^^^^^^^^^^^
KeyError: 'execution_count'
Shutting down kernel
thread 'Stdout reader' panicked at src\handlers\stream.rs:42:58:
not yet implemented: stream is dead
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'Stderr reader' panicked at src\handlers\stream.rs:42:58
```

**Comment by maxim-uvarov** [comment]
@cptpiepmatz I had no idea that all of this existed! Thanks a lot for showing me!!! 🙏🙏🙏

**Comment by maxim-uvarov** [comment]
@cptpiepmatz thanks a lot! Just to conclude without expecting any further assistance:

I used your references for further research and found that neither jupyter_console nor nb_client offers a CLI interface that would allow modifying and running single cells, nor partially executing notebooks and returning results. I mean, I thought of just sending a command `nb-client --connect some-core --execute '$a * $b'`, and receiving the output of this command, optionally adding this command as a new cell to some notebook.

All other proposed methods are out of my current technical capabilities.

Also, I found some projects that are more or less relevant to my goal:

- https://github.com/joouha/euporie
- https://github.com/mwouts/jupytext

Meanwhile I discovered how huge the Jupyter ecosystem is. I had no idea.

Though, while doing this research, I understood that I can just avoid creating and using variables, and if it is needed, save intermediate results to some files and, in the next steps, just open those results. This way, I don't need a running server, yet still have most of the functionality that I need\*.

\* well, `need` here is a loud word. It's more like - that I think would be interesting to experiment with:)

Thanks again!
